File: packages/kotlin-sdk/KotlinExampleApp/app/src/main/java/org/dashfoundation/example/ui/dashpay/ShieldedTipSubmission.kt
package org.dashfoundation.example.ui.dashpay

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/** Application-owned submissions survive sheet dismissal, navigation and activity recreation. */
class ShieldedTipSubmissions(private val scope: CoroutineScope) {
    private val wallets = mutableMapOf<String, ShieldedTipSubmission>()

    /** Called on the main thread; one guard per network/wallet, across all of its identities. */
    fun forWallet(network: Int, walletId: String): ShieldedTipSubmission =
        wallets.getOrPut("$network:$walletId") { ShieldedTipSubmission(scope) }
}

/** Main-thread state. A cancelled UI must never imply that blocking JNI stopped broadcasting. */
class ShieldedTipSubmission(private val scope: CoroutineScope) {
    enum class Status { Ready, Sending, Sent, Uncertain }

    var status by mutableStateOf(Status.Ready)
        private set
    var message by mutableStateOf<String?>(null)
        private set
    val busy: Boolean get() = status == Status.Sending
    val submitted: Boolean get() = status != Status.Ready

    fun submit(send: suspend () -> Unit) {
        if (submitted) return
        // Lock synchronously, before launching, so two UI events cannot submit twice.
        status = Status.Sending
        message = null
        scope.launch {
            try {
                send()
                status = Status.Sent
                message = "Shielded tip sent."
            } catch (error: Exception) {
                status = if (canReviewShieldedTipAfterFailure(error)) Status.Ready else Status.Uncertain
                message = if (status == Status.Uncertain) {
                    "The tip may have been sent. Check shielded activity before sending again. ${error.message.orEmpty()}"
                } else {
                    error.message ?: "Unable to send tip"
                }
                if (error is kotlinx.coroutines.CancellationException) throw error
            }
        }
    }

    /** Starting another payment requires an explicit action after a confirmed successful return. */
    fun startNewTip() {
        if (status != Status.Sent) return
        status = Status.Ready
        message = null
    }
}

File: packages/swift-sdk/SwiftExampleApp/SwiftExampleApp/Views/DashPay/ShieldedTipSubmission.swift
import Foundation
import SwiftDashSDK

/// Application-owned tip submissions. They outlive the sheet that started
/// them: a dismissed sheet must never imply the native send stopped
/// broadcasting, and reopening from either entry point (the DashPay tab or
/// the profile) has to find the same guard. Mirrors the Kotlin example app's
/// `ShieldedTipSubmissions`.
@MainActor
final class ShieldedTipSubmissions: ObservableObject {
    private var wallets: [String: ShieldedTipSubmission] = [:]

    /// One guard per network and wallet, across all of its identities.
    func forWallet(network: Network, walletId: Data) -> ShieldedTipSubmission {
        let key = "\(network.rawValue):\(walletId.hexString)"
        if let existing = wallets[key] { return existing }
        let created = ShieldedTipSubmission()
        wallets[key] = created
        return created
    }
}

/// Main-actor state of one wallet's shielded tip submission.
@MainActor
final class ShieldedTipSubmission: ObservableObject {
    enum Status: Equatable { case ready, sending, sent, uncertain }

    @Published private(set) var status: Status = .ready
    @Published private(set) var message: String?

    var busy: Bool { status == .sending }
    var submitted: Bool { status != .ready }

    /// Locks synchronously, before the task starts, so two UI events cannot
    /// submit twice. Returns nil when a submission already holds the lock; the
    /// task is returned so tests can await the outcome.
    @discardableResult
    func submit(_ send: @escaping @MainActor () async throws -> Void) -> Task<Void, Never>? {
        guard !submitted else { return nil }
        status = .sending
        message = nil
        return Task { @MainActor in
            do {
                try await send()
                status = .sent
                message = "Shielded tip sent."
            } catch {
                if Self.canReviewAfterFailure(error) {
                    status = .ready
                    message = error.localizedDescription
                } else {
                    status = .uncertain
                    message = "The tip may have been sent. Check shielded activity before sending "
                        + "again. \(error.localizedDescription)"
                }
            }
        }
    }

    /// Starting another payment is an explicit action after a confirmed success.
    func startNewTip() {
        guard status == .sent else { return }
        status = .ready
        message = nil
    }

    /// Only failures known not to have executed a tip permit a fresh review.
    /// On the tip path Rust maps selection, build and recipient-check failures
    /// to `walletOperation`, and broadcast ambiguity to
    /// `shieldedSpendUnconfirmed`; anything unknown, including cancellation,
    /// stays locked until shielded activity shows the outcome.
    static func canReviewAfterFailure(_ error: Error) -> Bool {
        guard let error = error as? PlatformWalletError else { return false }
        switch error {
        case .nullPointer, .invalidHandle, .invalidParameter, .invalidIdentifier, .invalidNetwork,
             .walletOperation, .identityNotFound, .contactNotFound, .utf8Conversion,
             .noSelectableInputs, .shieldedBroadcastFailed, .shieldedNoRecordedAnchor,
             .shieldedInsufficientBalance:
            return true
        default:
            return false
        }
    }
}
